src/string_id.h

```diff
@@ -69,69 +108,92 @@ template<typename T>
          */
-        string_id() : _cid( INVALID_CID ), _version( INVALID_VERSION ) {}
+        constexpr string_id() : string_id( "" ) {};
         /**
          * Comparison, only useful when the id is used in std::map or std::set as key. Compares
          * the string id as with the strings comparison.
          */
         bool operator<( const This &rhs ) const {
-            return _id < rhs._id;
+            // Not str() compares to prevent needless tier ups to std::string
+            return strcmp( c_str(), rhs.c_str() ) < 0;
         }
         /**
          * The usual comparator, compares the string id as usual.
          */
         bool operator==( const This &rhs ) const {
-            bool can_compare_cid = ( _cid != INVALID_CID || rhs._cid != INVALID_CID ) &&
-                                   _version == rhs._version && _version != INVALID_VERSION;
-            return ( can_compare_cid && _cid == rhs._cid ) ||
-                   ( !can_compare_cid && _id == rhs._id );
-            // else returns false, when:
-            //      can_compare_cid && _cid != rhs._cid
-            //  OR  !can_compare_cid && _id != rhs._id
+            return _hash == rhs._hash;
+        }
+        /**
+         * Allows to quickly compare hashes, when caching the hash would be faster
+         */
+        bool operator==( uint64_t hash ) const {
+            return _hash == hash;
         }
         /**
          * The usual comparator, compares the string id as usual.
          */
         bool operator!=( const This &rhs ) const {
-            return ! operator==( rhs );
+            return !operator==( rhs );
         }
         /**
          * Interface to the plain C-string of the id. This function mimics the std::string
          * object. Ids are often used in debug messages, where they are forwarded as C-strings
          * to be included in the format string, e.g. debugmsg("invalid id: %s", id.c_str())
          */
         const char *c_str() const {
-            return _id.c_str();
+            return _is_char_ptr ? _char_ptr : _string->c_str();
         }
         /**
          * Returns the identifier as plain std::string. Use with care, the plain string does not
          * have any information as what type of object it refers to (the T template parameter of
          * the class).
+         * Note: The function should be declared with safebuffers as MSVC insert a cookie check
+         * here for some reason, and it takes up 25% of str()'s runtime.
          */
         const std::string &str() const {
-            return _id;
-        }
-
-        explicit operator std::string() const {
-            return _id;
+            if( _is_char_ptr ) {
+                // We don't want to break the constness of str(), and this change is 100% internal with
+                // observable side effects, therefore we break the constness of this to cache the string.
```

Comment:
Why not just cache string literal ids as soon as the corresponding `string_id`s are constructed, so we won't need these two additional variables `_is_char_ptr` and `_char_ptr`? I doubt there are enough string literal ids in the c++ code for it to make any noticeable performance impact.

Reply:
std::strings cant be constructed in compile time, meaning loops and the such will still generate a new std::string every single time they are reached, even in loops.